### PR TITLE
feat: detect opengl version

### LINKED ISSUE #366: Add GPU Detection
Per https://github.com/FaithLife-Community/OuDedetai/issues/24#issuecomment-2618717194, the testing of ARM installations is currently held up because we don't have a machine that supports OpenGL 3.2 (ES?).

It would be good to have a test in OD that checks if the user has support for OpenGL 3.2 and if not, report that the install cannot proceed.

This should be a corner case for x86_64 installs, but it should be an easy check to introduce that the average user would never see except inside the log.

### PR DESCRIPTION
**Pull request by thw26**
Modified this to use a subprocess call rather than using the OpenGL Python libraries. This bypasses the GTK issues described below.

Fixes #366. Adds ability to get OpenGL version, and we compare the retrieved version to the minimum version.

Main test case is a PineTab2, which has a graphics limit of OpenGL 3.1. I'm not actually sure if OpenGL 3.2 is a minimum; it is just a guess based on my current knowledge of the lower limit.

We run this before install in order to prevent wasting the user's time.

We run this whenever we launch Logos to ensure that we can launch, e.g., if the user moves an install to a new computer.

Does some minor work in moving imports around in order to make the program run and to clean up ordering.

---

## TODO

- [x] Suppress error messages in TUI
  - The current implementation is making calls to Gtk which can cause Gtk and other system errors to be reported to terminal, and thus break the TUI. All my attempted implementations of blocking these have failed.

### CODE REVIEW

File: ou_dedetai/system.py [hunk 1]
```diff
@@ -366,6 +371,33 @@ class SuperuserCommandNotFound(Exception):
     """Superuser command not found. Install pkexec or sudo or doas"""
 
 
+def get_opengl_version(required_version="3.2"):
+    try:
+        result = subprocess.run(['glxinfo'], capture_output=True, text=True, check=True)
+    except FileNotFoundError:
+        return False, "glxinfo command not found. Please install mesa-utils or equivalent."
```
Comment (thw26) [inline ou_dedetai/system.py:378]: We perhaps need to modify our system packages here instead of providing this error.

File: ou_dedetai/tui_app.py [hunk 2]
```diff
@@ -1086,6 +1092,7 @@ def set_utilities_menu_options(self):
         )
         labels.append(label)
 
+        labels.append("Check OpenGL")
```
Comment (thw26) [inline ou_dedetai/tui_app.py:1095]: Added for quick testing, and it makes sense to have a utility to do so

File: ou_dedetai/system.py [hunk 3]
```diff
@@ -366,6 +371,33 @@ class SuperuserCommandNotFound(Exception):
     """Superuser command not found. Install pkexec or sudo or doas"""
 
 
+def get_opengl_version(required_version="3.2"):
+    try:
+        result = subprocess.run(['glxinfo'], capture_output=True, text=True, check=True)
```
[the remark this answers is not in the document]
Comment (ctrlaltf24) [reply ou_dedetai/system.py:376]: At very least this needs the LD_LIBRARY_PATH fix. See the invitation of subprocess in wine.py for how to use

**Review by ctrlaltf24 — CHANGES_REQUESTED** [review]

**Comment by ctrlaltf24** [comment]
Two main concerns:
1. Not sure if I love blocking the install process on this - seems more likely to do harm than good. Should be a warning instead? A logging.warning and app.status plus a sleep of say 5 seconds? So even if the output of glxinfo changes or what have you it still functions
2. mesa-utils wouldn't be installed for existing installs, nor do I see it in the system.py dependencies. We can't be relying on something we don't check to make sure it's installed. Plus OpenGL check is happing BEFORE we even install system dependencies. We'd need to move it to later

File: ou_dedetai/system.py [hunk 4]
```diff
@@ -366,6 +370,50 @@ class SuperuserCommandNotFound(Exception):
...
+        logging.info(message)
+        return True, message
+    else:
+        message = f"OpenGL Version: {opengl_version} is not supported (must be >= {required_version})."
+        logging.info(message)
+        return False, message
+
+
+class OpenGLIncompatible(Exception):
+    """Superuser command not found. Install pkexec or sudo or doas"""
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/system.py:398]: Hah! Nope. Copy pasta fail

File: ou_dedetai/tui_app.py [hunk 5]
```diff
@@ -772,6 +772,12 @@ def utilities_menu_select(self, choice):
                 self.reset_screen()
                 self.logos.switch_logging()
                 self.go_to_main_menu()
+            elif choice == "Check OpenGL":
+                self.reset_screen()
+                opengl_version, reason = system.get_opengl_version()
+                logging.info(reason)
+                self.status(reason)
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/tui_app.py:779]: No, thankfully. The console log displays "status messages" persistently.

File: ou_dedetai/installer.py [hunk 6]
```diff
@@ -271,11 +272,23 @@ def ensure_launcher_shortcuts(app: App):
...
+
+
 def install(app: App):
     """Entrypoint for installing"""
     app.status('Installing…')
     try:
+        ensure_opengl(app)
         ensure_launcher_shortcuts(app)
+    except OpenGLIncompatible:
+        app.status(f"Incompatible OpenGL version.")
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/installer.py:291]: Good observation. I think I coded it this way as I wasn't sure what to do with it at the time. I think, given your other feedback, we should perhaps have it disable installing and deliver a warning. The user won't be able to run Logos if OpenGL support is too low.

File: ou_dedetai/system.py (hunk 1 above)
Comment (thw26) [reply ou_dedetai/system.py:378]: In this case I meant modify the OS package lists in the our install command, or run an offer to package install to get it to work.

Comment (thw26) [reply ou_dedetai/system.py:378]: Submitting a commit that reflects my last comment.

File: ou_dedetai/system.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/system.py:398]: Submitting a commit to fix this soon.

**Comment by thw26** [comment]
> Two main concerns:
> 
>     1. Not sure if I love blocking the install process on this - seems more likely to do harm than good. Should be a warning instead? A logging.warning and app.status plus a sleep of say 5 seconds? So even if the output of glxinfo changes or what have you it still functions

My suggestion would be let's add a continue question?

>     2. mesa-utils wouldn't be installed for existing installs, nor do I see it in the system.py dependencies. We can't be relying on something we don't check to make sure it's installed. Plus OpenGL check is happing BEFORE we even install system dependencies. We'd need to move it to later

This I think is the main issue, even with the continue question. So, given this, the OpenGL check should happen right after package install, no?

**Review by ctrlaltf24 — CHANGES_REQUESTED** [review]

File: ou_dedetai/tui_app.py [hunk 7]
```diff
@@ -772,6 +772,12 @@ def utilities_menu_select(self, choice):
                 self.reset_screen()
                 self.logos.switch_logging()
                 self.go_to_main_menu()
+            elif choice == "Check OpenGL":
+                self.reset_screen()
+                opengl_version, reason = system.get_opengl_version(self)
+                logging.info(reason)
+                self.status(reason)
+                self.go_to_main_menu()
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/tui_app.py:780]: If the user hits a fault, it should ask the above approve. As my previous note remarks, something is wrong with how the TUI is handling ask/approve at present.

File: ou_dedetai/system.py [hunk 8]
```diff
@@ -485,6 +519,7 @@ def get_package_manager() -> PackageManager | None:
             "curl gawk grep "  # other
             "7zip cabextract "  # winetricks
             "xdg-utils " # For xdg-mime needed for custom url scheme registration
+            "Mesa-demo-x"  # verify opengl version
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/system.py:522]: What a strange name indeed.

File: ou_dedetai/system.py [hunk 9]
```diff
@@ -366,6 +370,34 @@ class SuperuserCommandNotFound(Exception):
     """Superuser command not found. Install pkexec or sudo or doas"""
 
 
+class OpenGLIncompatible(Exception):
+    """OpenGL version is incompatible."""
+
+
+def get_opengl_version(app: App, required_version="3.2"):
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/system.py:377]: Will fix.

File: ou_dedetai/system.py [hunk 10]
```diff
@@ -366,6 +370,34 @@ class SuperuserCommandNotFound(Exception):
...
+        env = wine.get_wine_env(app, None)
+        result = subprocess.run(['glxinfo'], env=fix_ld_library_path(env), capture_output=True, text=True, check=True)
+    except FileNotFoundError:
+        return False, "glxinfo command not found. Please install mesa-utils or equivalent."
+    except subprocess.CalledProcessError as e:
+        return False, f"glxinfo command failed: {e.stderr.strip()}"
+
+    match = re.search(r"OpenGL version string:\s+([\d\.]+)", result.stdout)
+    if not match:
+        return False, "Failed to parse OpenGL version from glxinfo output."
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/system.py:388]: Does this suggestion fit into #300?

File: ou_dedetai/logos.py [hunk 11]
```diff
@@ -118,6 +118,11 @@ def run_logos():
                 self.processes[self.app.conf.logos_exe] = process
             self.logos_state = State.RUNNING
 
+        # Ensure OpenGL version is compatible with Logos on Wine.
+        opengl_version, reason = system.get_opengl_version(self.app)
+        if not opengl_version:
+            self.app.exit(reason)
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai/logos.py:124]: If the OpenGL version is not running, Logos will fail to launch. Here I'm trying to catch that error in order to alert the user why Logos can't launch.

In current draft, this was an optional item (install even though we have a bad OGL version). I could produce the same ask/approve as above, perhaps? Tell user it won't run, run it, only to let it crash?

File: ou_dedetai.spec [hunk 12]
```diff
@@ -3,7 +3,7 @@
 
 a = Analysis(
     ['scripts/run_app.py'],
-    pathex=[],
+    pathex=['.'],
```
[the remark this answers is not in the document]
Comment (thw26) [reply ou_dedetai.spec:6]: I can test once more by removing this adjustment to see if I can retrigger the error and then share the specific error.

File: ou_dedetai/installer.py [hunk 13]
```diff
@@ -75,10 +76,31 @@ def ensure_sys_deps(app: App):
...
+        question = "Incompatible OpenGL version. Logos will be unable to launch. Should the install continue anyways?"
+        if app.approve(question):
+            logging.debug("> User continuing with incompatible OpenGL.")
+        else:
+            app.status("Exiting install…")
+            sys.exit()  # Exits the install thread.
+
+    logging.debug("> Done.")
```
[the remark this answers is not in the document]
Comment (ctrlaltf24) [reply ou_dedetai/installer.py:97]: `TUI wasn't properly handling approve() and auto assumed the first answer of yes`

the default install button sets assume_yes to True, bypassing all prompts - perhaps that was a poor design choice, but that's what happens. You can either store the value of assume yes before you prompt, force assume yes to false, then prompt, then restore the value of assume yes if you really want to prompt.

File: ou_dedetai/system.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ctrlaltf24) [reply ou_dedetai/system.py:388]: yup

File: ou_dedetai/logos.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (ctrlaltf24) [reply ou_dedetai/logos.py:124]: > In current draft, this was an optional item (install even though we have a bad OGL version). I could produce the same ask/approve as above, perhaps? Tell user it won't run, run it, only to let it crash?

IMO long term it'd be better to detect crashes and inform user rather than run ahead of time (in case open gl version changes or what have you). Don't like blocking Logos from starting, this can be fine for now, maybe with a fixme?

> I found the .exit implementation only kills the thread in this test rather than the whole app.

Seems like a good issue/general fix to get in (here or a different PR). Exiting app means you want to leave.

**Review by ctrlaltf24 — CHANGES_REQUESTED** [review]

**Review by ctrlaltf24 — APPROVED** [review]
